riscv/src/interrupt/supervisor.rs

```diff
@@ -0,0 +1,259 @@
+#[inline]
+pub unsafe fn enable() {
+    sstatus::set_sie()
+}
+
+/// Retrieves the cause of a trap in the current hart (supervisor mode).
+///
+/// This function expects the target-specific interrupt and exception types.
+/// If the raw cause is not a valid interrupt or exception for the target, it returns an error.
+#[inline]
+pub fn try_cause<I: CoreInterruptNumber, E: ExceptionNumber>() -> Result<Trap<I, E>> {
+    scause::read().cause().try_into()
+}
+
+/// Retrieves the cause of a trap in the current hart (supervisor mode).
+///
+/// This function expects the target-specific interrupt and exception types.
+/// If the raw cause is not a valid interrupt or exception for the target, it panics.
+#[inline]
+pub fn cause<I: CoreInterruptNumber, E: ExceptionNumber>() -> Trap<I, E> {
+    try_cause().unwrap()
+}
+
+/// Execute closure `f` with interrupts disabled in the current hart (supervisor mode).
+///
+/// This method does not synchronise multiple harts, so it is not suitable for
+/// using as a critical section. See the `critical-section` crate for a cross-platform
+/// way to enter a critical section which provides a `CriticalSection` token.
+///
+/// This crate provides an implementation for `critical-section` suitable for single-hart systems,
+/// based on disabling all interrupts. It can be enabled with the `critical-section-single-hart` feature.
+#[inline]
+pub fn free<F, R>(f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    let sstatus = sstatus::read();
+
+    // disable interrupts
+    disable();
+
+    let r = f();
+
+    // If the interrupts were active before our `disable` call, then re-enable
+    // them. Otherwise, keep them disabled
+    if sstatus.sie() {
+        unsafe { enable() };
+    }
+
+    r
+}
+
+/// Execute closure `f` with interrupts enabled in the current hart (supervisor mode).
+/// This method is assumed to be called within an interrupt handler, and allows
+/// nested interrupts to occur. After the closure `f` is executed, the [`sstatus`]
+/// and [`sepc`] registers are properly restored to their previous values.
+///
+/// # Safety
+///
+/// - Do not call this function inside a critical section.
+/// - This method is assumed to be called within an interrupt handler.
+/// - Make sure to clear the interrupt flag that caused the interrupt before calling
+///   this method. Otherwise, the interrupt will be re-triggered before executing `f`.
+#[inline]
+pub unsafe fn nested<F, R>(f: F) -> R
+where
+    F: FnOnce() -> R,
+{
+    let sstatus = sstatus::read();
+    let sepc = sepc::read();
+
+    // enable interrupts to allow nested interrupts
+    enable();
+
+    let r = f();
+
+    // If the interrupts were inactive before our `enable` call, then re-disable
+    // them. Otherwise, keep them enabled
+    if !sstatus.sie() {
+        disable();
+    }
+
+    // Restore SSTATUS.SPIE, SSTATUS.SPP, and SEPC
+    if sstatus.spie() {
+        sstatus::set_spie();
+    }
+    sstatus::set_spp(sstatus.spp());
```

Comment:
Similar suggestion to the machine interrupt handler.

Reply:
Same here, not sure about what is wrong with the current implementation

Reply:
> Same here, not sure about what is wrong with the current implementation

Nothing, on second inspection I saw the call to `csrrs`. I was mistaken before, and thought `csrrw` was called.